src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,259 @@
+     * message.
+     *
+     * The 16-bit `load_history` field is a ring buffer with a capacity for 4
+     * 4-bit entries.
+     *
+     * The 8-bit load field tracks an approximation of a cown's message queue
+     * depth as a count of messages processed between token messages falling out
+     * of the queue. When the token message falls out, the `reset_load()`
+     * function will push the upper nibble of the current load into the load
+     * history ring buffer.
+     */
+    uint32_t bits = 0 | needs_token_mask;
+
+  public:
+    /**
+     * Return true if this cown is muted.
+     */
+    inline bool muted() const
+    {
+      return (bits & muted_mask) != 0;
+    }
+
+    /**
+     * Return true if the token message is not in this cown's queue and a new
+     * token message should be added if it runs another message.
+     */
+    inline bool needs_token() const
+    {
+      return (bits & needs_token_mask) != 0;
+    }
+
+    /**
+     * Return the amount of messages processed since the last token message fell
+     * out of this cown's queue.
+     */
+    inline uint8_t current_load() const
+    {
+      return bits & current_load_mask;
+    }
+
+    /**
+     * Calculate the total load accumulated for this cown. This value will be
+     * in the range [0, 1215].
+     */
+    inline uint32_t total_load() const
+    {
+      // Add the current load to the 4 upper nibbles stored as load history. The
+      // load history may be more efficiently compressed, but the clang SLP
+      // vectorizer optimizes this nicely with AVX2 instructions.
+      const uint32_t h3 = (bits & 0x00'f000'00) >> 16;
+      const uint32_t h2 = (bits & 0x00'0f00'00) >> 12;
+      const uint32_t h1 = (bits & 0x00'00f0'00) >> 8;
+      const uint32_t h0 = (bits & 0x00'000f'00) >> 4;
+      return (h3 + h2 + h1 + h0) | current_load();
+    }
+
+    /**
+     * Return true if this cown is overloaded.
+     */
+    inline bool overloaded() const
+    {
+      return total_load() > overload_threshold;
+    }
+
+    /**
+     * Return true if this cown is muted or overloaded.
+     */
+    inline bool triggers_muting() const
+    {
+      return muted() || overloaded();
+    }
+
+    /**
+     * Return true if this cown is not muted and if the load is substantially
+     * lower than the threshold for muting.
+     */
+    inline bool triggers_unmuting() const
+    {
+      return !muted() && (total_load() < unmute_threshold);
+    }
+
+    /**
+     * Mark this cown as muted.
+     */
+    inline void mute()
+    {
+      assert(!muted());
+      bits |= muted_mask;
+    }
+
+    /**
+     * Unmark this cown as muted.
+     */
+    inline void unmute()
+    {
+      assert(muted());
+      bits &= ~muted_mask;
+    }
+
+    /**
+     * Mark that this cown has a token message in its queue and no longer
+     * requires a new token message to be added.
+     */
+    inline void add_token()
```

Comment:
the comment and the method name seems very confusing.

i would expect "add_token" to add a token.